Mage.Sets/src/mage/cards/c/CovetedFalcon.java

```diff
@@ -0,0 +1,152 @@
+        // Draw a card for each one they gained control of this way.
+        Ability turnedFaceUpTriggeredAbility = new TurnedFaceUpSourceTriggeredAbility(new CovetedFalconEffect());
+        turnedFaceUpTriggeredAbility.addTarget(new TargetOpponent());
+        turnedFaceUpTriggeredAbility.addTarget(
+                new TargetControlledPermanent(0, Integer.MAX_VALUE, StaticFilters.FILTER_CONTROLLED_PERMANENT, false));
+        this.addAbility(turnedFaceUpTriggeredAbility);
+    }
+
+    private CovetedFalcon(final CovetedFalcon card) {
+        super(card);
+    }
+
+    @Override
+    public CovetedFalcon copy() {
+        return new CovetedFalcon(this);
+    }
+}
+
+class CovetedFalconEffect extends ContinuousEffectImpl {
+
+    private boolean firstControlChange = true;
+    private List<UUID> permanentIds = new ArrayList<>();
+
+    CovetedFalconEffect() {
+        super(Duration.Custom, Layer.ControlChangingEffects_2, SubLayer.NA, Outcome.Benefit);
+        staticText = "target opponent gains control of any number of target permanents you control. " +
+                "Draw a card for each one they gained control of this way";
+    }
+
+    private CovetedFalconEffect(final CovetedFalconEffect effect) {
+        super(effect);
+        firstControlChange = effect.firstControlChange;
+        permanentIds = effect.permanentIds;
+    }
+
+    @Override
+    public boolean apply(Game game, Ability source) {
+        Player controller = game.getPlayer(source.getControllerId());
+        if (controller == null) {
+            discard(); // controller no longer exists
+            return false;
+        }
+
+        UUID controllingOpponentId = source.getFirstTarget();
+        if (firstControlChange) {
+            permanentIds.addAll(source.getTargets().get(1).getTargets());
+        }
+
+        int cardsToDraw = 0;
+        ListIterator<UUID> permanentsIterator = permanentIds.listIterator(permanentIds.size());
+        boolean oneTargetStillExists = false;
+        while (permanentsIterator.hasPrevious()) {
+            Permanent permanent = game.getPermanent(permanentsIterator.previous());
+            if (permanent == null) {
+                continue;
+            }
+            oneTargetStillExists = true;
+
+            if (permanent.isControlledBy(controllingOpponentId)) {
+                continue;
+            }
+
+            boolean controlChanged = permanent.changeControllerId(controllingOpponentId, game, source);
+            if (firstControlChange) {
+                if (controlChanged) {
+                    cardsToDraw++;
+                } else {
+                    // If we couldn't gain control of target permanent on the first try, we shouldn't try again
+                    // Can happen due to Guardian Beast
+                    permanentsIterator.remove();
+                }
+            }
+        }
+
+        if (cardsToDraw > 0) {
+            new DrawCardSourceControllerEffect(cardsToDraw).apply(game, source);
```

Comment:
I think applying OneShotEffect here is quite unsafe.
They may ask players choices/inputs (like if you have replacement effects to order on draw), and you really don't want to do that in the middle of applying continuous effects.

Maybe have a `OneShotEffect` that on `apply` does:
- keep only the targets that are still in battlefield not under the target player's control.
- adds control-change ContinuousEffect for all those targets.
- call game.getState().processAction(game);
- count each of targets MOR that are in control of opponent.

Reply:
Good point, I agree that it's better to have a OneShotEffect initiate continuous effects as needed, rather than a ContinuousEffect try to call OneShotEffects on first usage or init

Reply:
Oh, and another argument against it is it is applied in the middle of the layers.
So if you would draw your last card due to this, with a Laboratory Maniac enchanted by a Frogify on the battlefield, at layer 2 the Maniac has not yet lost his abilities, so you would win, which is obviously incorrect.

Reply:
I would advise having one `ContinuousEffect` per affected permanent as a workaround.